### ISSUE TITLE
[ENH] MicroPython support

### ISSUE THREAD

**Issue by kdschlosser** [cython/cython#3971]
**Is your feature request related to a problem? Please describe.**
Feature Request

With the ever growing world of IOT it would be nice to see Cython get used on the devices that run MicroPython.
MicroPython is a CPython port that runs on micro controllers/soc devices. It interfaces with the hardware available on these devices and has close to the full std libraries available as well as libraries that have been made for sensors that are able to be attached to the devices that MicroPython runs on. 

I don't see an immediate need to be able to bind to a C library, but what would be extremely useful is Cythons ability to convert Python code into C code and compile it as a C extension module. This would be really useful because of the hardware limitations of the devices. Python does carry with it a decent amount of overhead, this overhead is really seen when running Python on a resource handicapped device such as those used in IOT devices. Compiling python code into C extensions would give a pretty significant boost in performance to code running on one of these devices.

I believe that coupling MicroPython and Cython would really show the power of both applications. It would give the IOT world something really awesome by unlocking the potential of these devices in a way that even a "newbie" developer would be able to do.

**Comment by da-woods** [comment]
This looks relevant

https://github.com/micropython/micropython/issues/658

to me it looks like MicroPython's C API is sufficiently different from CPython's that it probably isn't realistic but it's possible I'm wrong in that assessment.

**Comment by kdschlosser** [comment]
yeah, there was discussion about it 6 years ago, it didn't get anywhere on the MicroPython end of things. So I thought I would bring it up  over here on the Cython end of things. 

It seems like it is possible, It would need to be a collaboration between the Cython devs and the MicroPython devs. Kind of like what was done to allow Cython to work properly with Stackless Python. I am also using a different point of view, I do not see a need to attach Python code to an existing C library, this is something that gets done on a full blown PC. There aren't any C libraries written so they would run on a micro controller in a manner that would be able to be used. I am looking at it from compiling python code into  a C extension only for the performance gains, which are going to be really substantial. Compiling python code into a C extension is really easy to do with Cython and it is something I have done a number of time in order to decrease code execution time.

**Comment by da-woods** [comment]
> Kind of like what was done to allow Cython to work properly with Stackless Python.

If I search the Cython utility code for "stackless" (case insensitive) I find a total of 6 lines (it's possible there were some significant changes on their end that I don't know about of course). It's genuinely a fork from CPython and so the C API compatibility is fairly close. In contrast MicroPython looks to be a from-scratch implementation of Python with a completely different C API. Pretty much every single line of utility code that we generate would need to be different.

> I do not see a need to attach Python code to an existing C library, this is something that gets done on a full blown PC. There aren't any C libraries written so they would run on a micro controller in a manner that would be able to be used. I am looking at it from compiling python code into a C extension only for the performance gains, which are going to be really substantial.

I think that wrapping C libraries would actually be the easy bit - it probably wouldn't need any changes at all. Compiling Python code to a completely new C API would be where all the work is.

--------------------------------------------

If someone was to do this I suspect the way forward would be:
1. Write a wrapper for the MicroPython C API that maps to the CPython limited API (since the limited API is designed to hide the implementation details of CPython, this would probably be easier than mapping to the regular CPython API). This wrapper would probably go into MicroPython rather than Cython.
2. Improve Cython's support for the limited API (which will probably happen anyway over time, but very basic code might work right now).

I'll stop commenting here, because I don't think I have much more to add.

**Comment by kdschlosser** [comment]
I don't know enough about the C end of CPython or MicroPython to be any help. IDK how much CPython there is left in MicroPython. It was mentioned in the issue posted earlier that mappings would need to be made so Cython would have an API to bind to.

All I am attempting to do is to get a dialog open between the 2 projects developers. I know that if a developer of an application comes to me and says hey. I have interest in making my program work with yours, I would spend some time to see if it is possible and what would need to be done. If a user comes along and says hey wouldn't it be nice if this program worked with yours I am going to be less inclined to make it happen if there is work that would need to be done on both ends.

This is kind of where I am stuck between a rock and a hard place because I am the user that sees a benefit to it. I am probably not explaining it in the best way on top of it. I know it had been discussed on the MicroPython end of things.

I am the person that came across the problems between Cython and Stackless. I went to the current maintainer of Stackless and I got lucky and I wasn't told "to bad it's not going to work" The developer worked with me in order to narrow down where the problem was. some of it was on the Cython end and some of it was on the Stackless side. The developer at that point opened a dialog with the Cython developers and it got taken care of.

An attempt was made to get the devs over at MicroPython to open a dialog with the folks over here... It didn't get any traction. 6 years later I thought it would give it another shot. I decided to attempt to do it the other way around and see if there is interest on this end and have the devs open a dialog with the folks over at MicroPython.

**Comment by robertwb** [comment]
Looking back at the conversation I participated in years ago and where things stand now, it looks like it might be possible to bridge a large portion of the gap between MicroPython and CPython's C API, but there would likely be significant differences when it comes to handling exceptions, and defining classes, etc. (including functionality that builds on that, such as closures and generators) that could probably only be bridged by special code and it's hard to get a handle on how significant of an effort that would be. 

I think that both the Cython and MicroPython developers would be happy to contribute ideas and brainstorm how to work around specific issues as they arise, but I imagine it would require finding someone else with sufficient time and motivation to step up and really drive this forward. @da-woods' comments on converging on the limited CPython API from both sides would probably be a good direction to start and increase the chances of overlapping with existing efforts.

**Comment by kdschlosser** [comment]
I wish I was proficient in C/CPP as this would be something I would be willing to do. Because of my current knowledge of C/CPP it would take a me horribly long time to do and I would also be a "newbie" pestering the seasoned developers in both Cython and MicroPython. Someone proficient in C/CPP is needed and ideally they would need to be familiar with the C code in MicroPython and in Cython.

**Comment by da-woods** [comment]
Probably worth adding: all the code that Cython currently generates is targetting variations of the CPython API. We support PyPy because it's implemented that API (and thus Cython only needs minor variations).

You might want to look into HPy - it's aiming to be a "better designed" C API that's less tied to the CPython implementation. There  are (early) plans to have Cython generate HPy code. MicroPython would probably find it easier to work with HPy than to implement the CPython API. If MicroPython supported HPy and Cython supported HPy it'd be pretty close to working.

Very little of that's implemented right now though.

> If I were to implement it myself, where would I start looking in the code? Or where can I find documentation about the structure and organization of Cython's implementation?

If you want to start looking at the code generation side there's two places I'd start:
1. the utility code in Cython/Utility/*.c - you'd need to reimplement most of it
2. In Nodes.py and ExprNodes.py - the functions starting with "generate_". the classes ExprNode and Node have reasonable summaries of the process in their comments

That doesn't cover everything, but it's where I'd start trying to understand it.

**Comment by KeithCu** [comment]
What about making a header file which does a mapping for Cython to MicroPython? It would be great to run compiled code on MicroPython. 

Here's an example one. Obviously it would need to get a lot larger, but it's pretty boilerplate, assuming it works. It might be always incomplete, but just getting typical embedded programs compiled with Cython and working with MicroPython would be huge.

[micropython.txt](https://github.com/user-attachments/files/18927360/micropython.txt)

**Comment by kdschlosser** [comment]
I have actually spent a great deal of time learning the MicroPython C API. I have to say what @KeithCu has linked to is really good. There is going to be some other goofy kind of stuff that will need to be dealt with specifically creating python functions and things of that nature but the basics stuff is there in that file. 

That's awesome that you or someone spent the time to key that up.

**Comment by kdschlosser** [comment]
It would be a huge deal for Cython to be able to convert something like Python code into the C Code needed to run on a microcontroller. It sure would open up the doors for potential uses.

There is a project called the lvgl micropython binding and in that project there is a script that was written to read the header files of LVGL and it then writs the C code needed to make LVGL accessible from inside of MicroPython. Cool thing is it works on more than just LVGL. It can be used for a lot of other C libraries so long as they can be compiled using the MCU's compiler. c99 is pretty much a given across the board for almost all MCU's. A lot of them can use newer versions of the C/CPP standard.

Cool thing about that LVGL MicroPython binding is the ability to compile MicroPython on nix type OS's so you end up with a nice easy to use GUI framework in less than 3mb of drive space. The memory use is very small as well. a 640 x 480 x 16bpp can be done with as little as 1mb or ram. 

MicroPython is not just limited to running on an MCU.. Check this out...

https://sim.lvgl.io/v8.3/micropython/ports/javascript/index.html

That is compiled to run as a web application.

**Comment by KeithCu** [comment]
Actually it was just a bit of coaxing with Grok 3! I'm planning on keeping working on it to eventually get (some) Cython working with MicroPython, but I'm sure other people with actual knowledge could do it faster.

**Comment by kdschlosser** [comment]
If you create a repo I can help out with it if ya want...

**Comment by KeithCu** [comment]
I created a repo: https://github.com/KeithCu/Cython2Upy I'll work on it in my free time. Kevin, I invited you to have write access.
